File: bubble_sort.rb
# Сортировка пузырьком на Ruby

def bubble_sort(array)
  n = array.length
  loop do
    swapped = false

    (n - 1).times do |i|
      if array[i] > array[i + 1]
        array[i], array[i + 1] = array[i + 1], array[i]
        swapped = true
      end
    end

    break unless swapped
  end

  array
end

# Пример использования:
if __FILE__ == $PROGRAM_NAME
  arr = [5, 3, 8, 4, 2]
  puts "Исходный массив: #{arr.inspect}"
  sorted_arr = bubble_sort(arr.dup)
  puts "Отсортированный массив: #{sorted_arr.inspect}"
end

File: BubbleSort.java
public class BubbleSort {

    public static void bubbleSort(int[] arr) {
        int n = arr.length;
        boolean swapped;
        
        for (int i = 0; i < n - 1; i++) {
            swapped = false;
            
            // Последние i элементов уже отсортированы
            for (int j = 0; j < n - i - 1; j++) {
                if (arr[j] > arr[j + 1]) {
                    // Меняем местами элементы
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                    swapped = true;
                }
            }
            
            // Если не было обменов, массив уже отсортирован
            if (!swapped) {
                break;
            }
        }
    }

    public static void main(String[] args) {
        int[] array = {64, 34, 25, 12, 22, 11, 90};
        
        System.out.println("Исходный массив:");
        printArray(array);
        
        bubbleSort(array);
        
        System.out.println("Отсортированный массив:");
        printArray(array);
    }
    
    public static void printArray(int[] arr) {
        for (int value : arr) {
            System.out.print(value + " ");
        }
        System.out.println();
    }
}
